csharp/src/Google.Protobuf/JsonTokenizer.cs

```diff
@@ -47,32 +47,38 @@ namespace Google.Protobuf
     /// between values. It validates the token stream as it goes - so callers can assume that the
     /// tokens it produces are appropriate. For example, it would never produce "start object, end array."
     /// </para>
+    /// <para>Implementation details: the base class handles single token push-back and </para>
     /// <para>Not thread-safe.</para>
     /// </remarks>
-    internal sealed class JsonTokenizer
+    internal abstract class JsonTokenizer
```

Comment:
An alternative class design would be to  provide push-back functionality as a decorator on top of a basic tokenizer (same idea as BufferedReader wrapping Reader). Since this is not part of public API, it doesn't matter much and the current approach is alright.